Data Docs 2.0

saladoc opened the pull request:
The first section of the data docs is aimed at people using it for the first time, so explanations may seem overly verbose. It is not my intent to explore every detail and every possiblitity of the Data API but to give an overview and an introduction which enables the user to utilise features relying on the Data API (Items, Blocks, Entities, ...).

This PR is open to feedback regarding the content and code examples; the wording may be subject to change, though suggestions are also welcome. 

Last but not least, feel free to list any topics not yet included which in your opinion seem important to have for beginner's docs. 
- [X] Explain how to use keys
- [X] Explain `DataManipulator` usage and use cases
- [x] `DataTransactionResult`

Commit:
Begin Data API docs 2.0

 - Overview / Glossary
 - Token API

Commit:
Flesh out data manipulators, create stub for serialization page.

Spongy commented:
Setting up PR reference, please wait...

gabizou commented on source/plugin/data/datamanipulators.rst:
```diff
@@ -0,0 +1,85 @@
+
+    public void swapHealth(DataHolder targetA, DataHolder targetB) {
+        if (targetA.supports(HealthData.class) && targetB.supports(HealthData.class)) {
+            HealthData healthA = targetA.get(HealthData.class).orNull();
```
If you have the guarantee, use `get()` instead of `orNull()`.

Commit:
Use .get() instead of .orNull()

Commit:
Fix data manipulator code example + small fixes + transactions page stub

Commit:
Move value containers to their own section

Commit:
Data Transactions

Commit:
Enforce char limit per line

Commit:
Do not include serialization at this point

saladoc commented:
Since this now explains the core features of the Data API, it is best gotten out as soon as possible. Therefore I'll open it up for review and nitpicking, additional stuff (like Serialization) can be included later.

@gabizou Go wild, if you have the time

@ the rest, please help me make sure that these docs are as understandable as they need to be

RobertHerhold commented on source/plugin/data/index.rst:
```diff
@@ -0,0 +1,78 @@
+Property
+~~~~~~~~
+
+A property is a read-only point of data. It provides information that is directly derived from its holders type,
```
"holders" --> "holder's"

RobertHerhold commented on source/plugin/data/index.rst:
```diff
@@ -0,0 +1,78 @@
+~~~~~~~~
+
+A property is a read-only point of data. It provides information that is directly derived from its holders type,
+but, unlike other data, unchangeable (except for core mods). Example for this are the harvesting ablities on tools
```
I would add "it is" before unchangeable

RobertHerhold replied:
"Example for this are" --> "Examples of this include"

RobertHerhold commented on source/plugin/data/index.rst:
```diff
@@ -0,0 +1,78 @@
+A data manipulator represents points of cohesive data that describes a certain component of its holder. For
+example ``HealthData``, which contains both current and maximum health. If a data holder has ``HealthData``, it
+has health that can somehow be depleted and replenished and can die if that health is depleted. This allows for
+the re-use of such components over the API and prevents duplication of accessor methods - as sheep, stained glass
```
Starting with "... - as sheep..." is worded a bit weirdly... Maybe put a period after "methods" and say "For example, sheep, stained blocks, and leather armor..."

RobertHerhold commented on source/plugin/data/keys.rst:
```diff
@@ -0,0 +1,109 @@
+
+Now for the details of the above function.
+
+The first line checks if our given data holder possesses health. Only if he does, he can be healed after all.
```
"he" --> "it"

RobertHerhold commented on source/plugin/data/keys.rst:
```diff
@@ -0,0 +1,109 @@
+Now for the details of the above function.
+
+The first line checks if our given data holder possesses health. Only if he does, he can be healed after all.
+Since a data holder can not have health without having a maximum health and vice versa, a check for one of the
```
Not sure that this logic makes sense... Maybe "Since any particular data holder is not aware of it's data types, check if it the key is applicable to the data holder using ``supports()``."

saladoc replied:
Basically, since `HEALTH` and `MAX_HEALTH` are treated by the same data processor (as they are part of the same data manipulator), checking if one of them is supported implies the other is too.
Are you saying I should check for both keys, and for the sake of this docs page act like all keys are independent from each other?

RobertHerhold commented on source/plugin/data/keys.rst:
```diff
@@ -0,0 +1,109 @@
+
+The second line uses the ``getOrNull()`` function to ask the data holder for its maximum health. Besides
+``getOrNull()``, the methods ``get()`` and ``getOrElse()`` exist, all of which accept a ``Key`` as their first
+parameter. Generally, ``get()`` should be used. It will return an ``Optional`` of the data requested or
```
"Generally, ``get()`` should be used. It..." --> "Generally, ``get()`` should be used, which will..."

RobertHerhold commented on source/plugin/data/keys.rst:
```diff
@@ -0,0 +1,109 @@
+Transforming Data
+=================
+
+Other than getting a value, modify it and offer it back, there is another way of modifying data. Using a data
```
"Other than getting a value, modify it and offer it back" --> "Other than getting, modifying, and offering a value"

RobertHerhold replied:
"modifiying" --> "interacting with"

RobertHerhold commented on source/plugin/data/keys.rst:
```diff
@@ -0,0 +1,109 @@
+=================
+
+Other than getting a value, modify it and offer it back, there is another way of modifying data. Using a data
+holders ``transform()`` method we can pass a ``Key`` and a ``Function``. Internally, the value for the key will
```
"holders" --> "holder's"

RobertHerhold commented on source/plugin/data/keys.rst:
```diff
@@ -0,0 +1,109 @@
+
+Other than getting a value, modify it and offer it back, there is another way of modifying data. Using a data
+holders ``transform()`` method we can pass a ``Key`` and a ``Function``. Internally, the value for the key will
+retrieved and the given function applied to it. The result is then stored under the key and the ``transform()``
```
"retrieved" --> "be retrieved"

RobertHerhold commented on source/plugin/data/datamanipulators.rst:
```diff
@@ -0,0 +1,111 @@
+data by passing its class to the ``getOrCreate()`` method. We get an ``Optional`` which we can use for our check.
+If the target does not support health data, it will be absent. But if the health data is present, it now contains
+a mutable copy of the data present on the data holder. We make our alterations and finally offer the changed data
+back to our target, where it is accepted (again, ``offer`` will return a ``DataTransactionResult`` which we will
```
"``offer``" --> "``offer()``"

RobertHerhold commented on source/plugin/data/datamanipulators.rst:
```diff
@@ -0,0 +1,111 @@
+Mutable vs. Immutable Data Manipulators
+=======================================
+
+To every data manipulator, there is a matching ``ImmutableDataHolder``. For instance both ``HealthData`` and
```
comma after "instance"

RobertHerhold commented on source/plugin/data/datamanipulators.rst:
```diff
@@ -0,0 +1,111 @@
+=======================================
+
+To every data manipulator, there is a matching ``ImmutableDataHolder``. For instance both ``HealthData`` and
+``ImmutableHealthData`` contain the same data, only the latter does not provide any means to make alterations to
```
"only" --> "but" (just sounds better to me, you may disagree)

RobertHerhold commented on source/plugin/data/datamanipulators.rst:
```diff
@@ -0,0 +1,111 @@
+
+Conversion between mutable and immutable data manipulators is done via the ``asImmutable()`` and ``asMutable()``
+methods, which each will return a copy of the data. Since the only way to obtain an immutable data manipulator
+from a data holder is obtaining a mutable one and then using ``asImmutable()`` in terms of processing power it
```
comma after ``asImmutable()``

RobertHerhold commented on source/plugin/data/transactions.rst:
```diff
@@ -0,0 +1,90 @@
+| ``CANCELLED`` | An event for this transaction was cancelled                                |
++---------------+----------------------------------------------------------------------------+
+
+A ton of immutable data
```
Not sure this heading is needed? Not sure what it's for.

saladoc replied:
I actually left it as a placeholder and then forgot to change it...

RobertHerhold commented on source/plugin/data/transactions.rst:
```diff
@@ -0,0 +1,90 @@
+Offering HealthData to a block of stone
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Now our above-mentioned examples are coded in a Way that they will fail silently rather than try to offer the
```
"Way" --> "way"

Saladoc committed:
Scrutiny fixes

 - Fixed various small problems pointed out by @RobertHerhold
 - use get() instead of getOrNull() for consistency
 - consistency in Code Example heading

gabizou commented on source/plugin/data/datamanipulators.rst:
```diff
@@ -0,0 +1,111 @@
+First we need to check if our target has health data. We do so by first asking it to provide us with its health
+data by passing its class to the ``getOrCreate()`` method. We get an ``Optional`` which we can use for our check.
+If the target does not support health data, it will be absent. But if the health data is present, it now contains
+a mutable copy of the data present on the data holder. We make our alterations and finally offer the changed data
```
Don't forget to include that if the data is "supported" but not available, it is still created. (the whole point of `getOrCreate()` versus `get()`).

gabizou commented on source/plugin/data/datamanipulators.rst:
```diff
@@ -0,0 +1,111 @@
+If the target does not support health data, it will be absent. But if the health data is present, it now contains
+a mutable copy of the data present on the data holder. We make our alterations and finally offer the changed data
+back to our target, where it is accepted (again, ``offer()`` will return a ``DataTransactionResult`` which we will
+just discard here).
```
Not discard, but rather, disregarding. Even though, sometimes it should be checked for any possible errors.

gabizou commented on source/plugin/data/datamanipulators.rst:
```diff
@@ -0,0 +1,111 @@
+
+    public void swapHealth(DataHolder targetA, DataHolder targetB) {
+        if (targetA.supports(HealthData.class) && targetB.supports(HealthData.class)) {
+            HealthData healthA = targetA.get(HealthData.class).get();
```
Again, emphasis on the difference between `get()` and `getOrCreate()`.

gabizou commented on source/plugin/data/datamanipulators.rst:
```diff
@@ -0,0 +1,111 @@
+
+.. code-block:: java
+
+    public  <T extends DataManipulator<?,?>> void swapData(DataHolder targetA, DataHolder targetB, Class<T> dataClass) {
```
I'm not sure if this will compile with generics, have you checked?

gabizou commented on source/plugin/data/datamanipulators.rst:
```diff
@@ -0,0 +1,111 @@
+Mutable vs. Immutable Data Manipulators
+=======================================
+
+To every data manipulator, there is a matching ``ImmutableDataHolder``. For instance, both ``HealthData`` and
```
I'm sure you meant `ImmutableDataManipulator` not `ImmutableDataHolder`.

gabizou commented on source/plugin/data/datamanipulators.rst:
```diff
@@ -0,0 +1,111 @@
+=======================================
+
+To every data manipulator, there is a matching ``ImmutableDataHolder``. For instance, both ``HealthData`` and
+``ImmutableHealthData`` contain the same data, only the latter does not provide any means to make alterations to
```
"only the latter returns new instances when requesting modified data."

gabizou commented on source/plugin/data/datamanipulators.rst:
```diff
@@ -0,0 +1,111 @@
+
+Conversion between mutable and immutable data manipulators is done via the ``asImmutable()`` and ``asMutable()``
+methods, which each will return a copy of the data. Since the only way to obtain an immutable data manipulator
+from a data holder is obtaining a mutable one and then using ``asImmutable()``, in terms of processing power it
```
You don't really need to mention processing power here. Object creation/lookups is really quite cheap in Java.

gabizou commented on source/plugin/data/index.rst:
```diff
@@ -0,0 +1,78 @@
+Overview
+========
+
+The unified Data API aims to provide a consistent way of accessing and modifying data of any kind. Where other
```
Brain dump on defining "data":

#### Data
Any type of data that is consistently synchronized from the game server to the clients. If it can be changed and is synchronized to the connected clients, that is Data that the Data API should cover. Properties on the other hand, are data that can *not* be synchronized from the server to the client throughout the lifetime of the game, without requiring modifications to both the server and the client. Properties are explained in "insert article on properties here".

gabizou commented on source/plugin/data/index.rst:
```diff
@@ -0,0 +1,78 @@
+interface itself is completely oblivious to the type of data held. Since only the implementations will possess
+this knowledge, it is possible to ask a ``DataHolder`` to provide data it does not have or to accept data it
+cannot use. In those cases, the return values of the methods will provide the information that data is not
+available (via ``Optional.absent()``) or not accepted.
```
or `DataTransactionResult.Type.FAILURE`.

gabizou commented on source/plugin/data/index.rst:
```diff
@@ -0,0 +1,78 @@
+On first glance at the API docs, the data API threatens to overwhelm you with lots of interfaces and packages. But
+to simply use the data API, you will not have to deal with many of them, as most interfaces found there are just
+specific data manipulators.
+
```
Should likely also include the notion of `ValueContainer`s that are the backend of Keys.

saladoc replied:
If they are the backend, should they really get included into the basic section? I tried to include only those which a plugin developer will directly interact with. Are `ValueContainer`s relevant for basic users or should they rather be spared until documentation for advanced users follows?

gabizou commented on source/plugin/data/index.rst:
```diff
@@ -0,0 +1,78 @@
+~~~~~
+
+Within the Data API, a value referred to by a ``Key`` is encoded in a container object. For this documentation, it
+is referred to as 'value container' to avoid confusion with the actual value, but note that the class name
```
Should likely find a different word to really avoid confusion as to which `ValueContainer` you're referring to. Maybe a "keyed value" to represent a `Value`.

gabizou commented on source/plugin/data/keys.rst:
```diff
@@ -0,0 +1,107 @@
+all. Since a data holder can not have current health without having a maximum health and vice versa, a check for
+one of the keys using the ``supports()`` method suffices.
+
+The second line uses the ``getOrNull()`` function to ask the data holder for its maximum health. Besides
```
you mean `get()` since `getOrNull()` was not used.

gabizou commented on source/plugin/data/keys.rst:
```diff
@@ -0,0 +1,107 @@
+        });
+    }
+
+Or, since we use Java 8 and are able to make use of its lambda expressions:
```
"Or if you use Java 8, you're able to shorten the line with lambda expressions:"

Saladoc committed:
Dat data tho

 - worked braindump into description of data and properties
 - made descriptions of code examples match the code and vice versa
 - 'value container' -> 'keyed value'
 - several small fixes based on gabizous suggestions

Saladoc committed:
character limit per line

saladoc commented:
Squashe and merged in 4dda74cf7ad48734cc8f968c4207b044494a7313
